connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala

```diff
@@ -0,0 +1,213 @@
+          (!ignoreNullable || !sqlField.nullable)) {
+          if (positionalFieldMatch) {
+            throw new IncompatibleSchemaException("Cannot find field at position " +
+              s"$sqlPos of ${toFieldStr(protoPath)} from Protobuf schema " +
+              s"(using positional matching)")
+          } else {
+            throw new IncompatibleSchemaException(
+              s"Cannot find ${toFieldStr(catalystPath :+ sqlField.name)} in Protobuf schema")
+          }
+        }
+      }
+
+    /**
+     * Validate that there are no Protobuf fields which don't have a matching Catalyst field,
+     * throwing [[IncompatibleSchemaException]] if such extra fields are found.
+     * Only required (non-nullable) fields are checked; nullable fields are ignored.
+     */
+    def validateNoExtraRequiredProtoFields(): Unit = {
+      val extraFields = protoFieldArray.toSet -- matchedFields.map(_.fieldDescriptor)
+      extraFields.filterNot(isNullable).foreach { extraField =>
+        if (positionalFieldMatch) {
+          throw new IncompatibleSchemaException(s"Found field '${extraField.getName()}'" +
+            s" at position ${extraField.getIndex} of ${toFieldStr(protoPath)} from Protobuf " +
+            s"schema but there is no match in the SQL schema at ${toFieldStr(catalystPath)} " +
+            s"(using positional matching)")
+        } else {
+          throw new IncompatibleSchemaException(
+            s"Found ${toFieldStr(protoPath :+ extraField.getName())} in Protobuf schema " +
+              "but there is no match in the SQL schema")
+        }
+      }
+    }
+
+    /**
+     * Extract a single field from the contained Protobuf schema which has the desired field name,
+     * performing the matching with proper case sensitivity according to SQLConf.resolver.
+     *
+     * @param name The name of the field to search for.
+     * @return `Some(match)` if a matching Protobuf field is found, otherwise `None`.
+     */
+    private[protobuf] def getFieldByName(name: String): Option[FieldDescriptor] = {
+
+      // get candidates, ignoring case of field name
+      val candidates = fieldMap.getOrElse(name.toLowerCase(Locale.ROOT), Seq.empty)
+
+      // search candidates, taking into account case sensitivity settings
+      candidates.filter(f => SQLConf.get.resolver(f.getName(), name)) match {
+        case Seq(protoField) => Some(protoField)
+        case Seq() => None
+        case matches => throw new IncompatibleSchemaException(s"Searching for '$name' in " +
+          s"Protobuf schema at ${toFieldStr(protoPath)} gave ${matches.size} matches. " +
+          s"Candidates: " + matches.map(_.getName()).mkString("[", ", ", "]")
+        )
+      }
+    }
+
+    /** Get the Protobuf field corresponding to the provided Catalyst field name/position,
+     *  if any. */
+    def getProtoField(fieldName: String, catalystPos: Int): Option[FieldDescriptor] = {
+      if (positionalFieldMatch) {
```

Comment:
Do we support `positionFieldMatch`?  I think we can remove it. It will be tricky to do and unncessary since Protobufs have field numbers.

Reply:
Yeah, this should be removed